service/src/main/java/bio/terra/pipelines/service/PipelineInputsOutputsService.java

```diff
@@ -114,6 +123,241 @@ public Map<String, Object> retrievePipelineInputs(PipelineRun pipelineRun) {
+   *
+   * <p>Note that this method does not perform any validation on the inputs. We expect the inputs to
+   * have been validated (e.g. all required inputs are present and of the correct type).
+   *
+   * <p>This does not cast the inputs to the correct type, nor does it format any file inputs with
+   * storage container URLs.
+   *
+   * @param allInputDefinitions - all the input definitions for a pipeline
+   * @param userProvidedPipelineInputs - the user-provided inputs
+   * @return Map<String, Object> allPipelineInputs - the combined inputs
+   */
+  public Map<String, Object> gatherRawInputs(
+      List<PipelineInputDefinition> allInputDefinitions,
+      Map<String, Object> userProvidedPipelineInputs) {
+
+    Map<String, Object> rawInputs = new HashMap<>();
+
+    allInputDefinitions.forEach(
+        inputDefinition -> {
+          String inputName = inputDefinition.getName();
+          if (inputDefinition.isUserProvided()) {
+            rawInputs.put(
+                inputName,
+                userProvidedPipelineInputs.getOrDefault(
+                    inputName, inputDefinition.getDefaultValue()));
+          } else {
+            rawInputs.put(inputName, inputDefinition.getDefaultValue());
+          }
+        });
+
+    logger.info("Populated raw pipeline inputs: {}", rawInputs);
+
+    return rawInputs;
+  }
+
+  /**
+   * Format the pipeline inputs for a pipeline. Apply the following manipulations: - use custom
+   * (environment-specific) values for certain service-provided inputs - prepend the storage
+   * workspace container URL to the service-provided inputs that need it - prepend the control
+   * workspace container URL to the user-provided file inputs - cast all the inputs according to the
+   * type specified in the pipeline input definitions
+   */
+  public Map<String, Object> formatPipelineInputs(
+      Map<String, Object> allRawInputs,
+      List<PipelineInputDefinition> allInputDefinitions,
+      UUID jobId,
+      String controlWorkspaceContainerUrl,
+      Map<String, Object> inputsWithCustomValues,
+      List<String> keysToPrependWithStorageWorkspaceContainerUrl,
+      String storageWorkspaceContainerUrl) {
+    Map<String, Object> formattedPipelineInputs = new HashMap<>();
+
+    for (PipelineInputDefinition inputDefinition : allInputDefinitions) {
+      String keyName = inputDefinition.getName();
+      String wdlVariableName = inputDefinition.getWdlVariableName();
+      PipelineVariableTypesEnum pipelineInputType = inputDefinition.getType();
+
+      // use custom value if present, otherwise use the value from raw inputs (allRawInputs)
+      String rawValue =
+          inputsWithCustomValues.getOrDefault(keyName, allRawInputs.get(keyName)).toString();
+      String processedValue;
+
+      if (keysToPrependWithStorageWorkspaceContainerUrl.contains(keyName)) {
+        // the rawValue for this field should start with a / so we don't need to add one here
```

Comment:
we don't really check this anywhere, not sure if we should add more logic for this

Reply:
Does GCP not care if you have an extra slash?

Reply:
nope you are totally right it does, i was thinking of http requests not stupid object storage paths

Reply:
added a FileUtils method to take care of this